File: frontend/services/AccountingEngine.ts
import { CardModel, DeclaredProject, ProjectType, Suit, PlayerPosition, ScoreBreakdown, Rank, DoublingLevel } from '../types';
import { POINT_VALUES } from '../utils/gameLogic';

/**
 * ACCOUNTING ENGINE
 * 
 * Based on comprehensive research of the standard scoring system.
 * 
 * KEY CONSTANTS:
 * - SUN Total: 130 Abnat (120 cards + 10 last trick) → 26 Game Points
 * - HOKUM Total: 162 Abnat (152 cards + 10 last trick) → 16 Game Points
 * 
 * FORMULAS:
 * - SUN: Points = (Abnat × 2) / 10  (no rounding needed - always whole number)
 * - HOKUM: Points = Round(Abnat / 10) with 0.5 rounding DOWN
 * 
 * WIN CONDITIONS:
 * - SUN: Buyer must score > 13 points (> 65 Abnat)
 * - HOKUM: Buyer must score > 8 points
 * 
 * KHASARA (Buyer Loss):
 * - SUN: Opponent takes ALL 26 points
 * - HOKUM: Opponent takes ALL 16 points
 * 
 * KABOOT (Capot - All Tricks):
 * - SUN: 44 points (includes 10-point last trick bonus doubled: 120*2/10 + 2*2 = 26 + 8... actually just fixed 44)
 * - HOKUM: 25 points
 */
export class AccountingEngine {

    /**
     * Calculates the final Game Points for a round given the raw inputs.
     * This is the "Truth" source for scoring - matches standard rules exactly.
     */
    public static calculateRoundResult(
        usRaw: number,           // Raw Abnat for "Us" team (includes last trick bonus if won)
        themRaw: number,         // Raw Abnat for "Them" team
        usProjects: number,      // Sum of project Abnat values (NOT game points yet)
        themProjects: number,
        bidType: 'SUN' | 'HOKUM',
        doublingLevel: DoublingLevel,
        bidderTeam: 'us' | 'them' | null,
        hasBalootUs: boolean = false,    // NEW: Track Baloot separately (2 pts, never multiplied)
        hasBalootThem: boolean = false
    ): { us: ScoreBreakdown, them: ScoreBreakdown, winner: 'us' | 'them' | 'tie', baida: boolean } {

        // ═══════════════════════════════════════════════════════════
        // 1. INITIAL STRUCTURES
        // ═══════════════════════════════════════════════════════════
        const us: ScoreBreakdown = {
            rawCardPoints: usRaw,
            projectPoints: usProjects,
            totalRaw: usRaw + usProjects,
            gamePoints: 0,
            isKaboot: false,
            multiplierApplied: doublingLevel
        };

        const them: ScoreBreakdown = {
            rawCardPoints: themRaw,
            projectPoints: themProjects,
            totalRaw: themRaw + themProjects,
            gamePoints: 0,
            isKaboot: false,
            multiplierApplied: doublingLevel
        };

        // ═══════════════════════════════════════════════════════════
        // 2. KABOOT DETECTION (All Tricks Won)
        // ═══════════════════════════════════════════════════════════
        // If opponent scored 0 raw Abnat, it's a Kaboot (they got no tricks)
        const isUsKaboot = themRaw === 0 && usRaw > 0;
        const isThemKaboot = usRaw === 0 && themRaw > 0;

        // ═══════════════════════════════════════════════════════════
        // 3. CONVERSION FUNCTIONS (Abnat → Game Points)
        // ═══════════════════════════════════════════════════════════

        /**
         * SUN CONVERSION: Points = (Abnat × 2) / 10
         * - This formula naturally handles .5 cases (65 × 2 = 130, 130/10 = 13)
         * - Total pool: 130 Abnat → 26 Game Points
         */
        const convertSunAbnat = (abnat: number): number => {
            return (abnat * 2) / 10;
        };

        /**
         * HOKUM CONVERSION: Points = Round(Abnat / 10) with 0.5 rounding DOWN
         * - 15.5 → 15 (round half down per ExternalApp)
         * - 15.6 → 16
         * - Total pool: 162 Abnat → 16 Game Points
         */
        const convertHokumAbnat = (abnat: number): number => {
            const divided = abnat / 10;
            const decimal = divided - Math.floor(divided);

            // Round half DOWN (Standard rule: 15.5 → 15)
            if (decimal <= 0.5) {
                return Math.floor(divided);
            } else {
                return Math.ceil(divided);
            }
        };

        // ═══════════════════════════════════════════════════════════
        // 4. BASE CALCULATION
        // ═══════════════════════════════════════════════════════════
        let usBase = 0;
        let themBase = 0;
        const maxPoints = bidType === 'SUN' ? 26 : 16;

        if (isUsKaboot) {
            // US won all tricks
            us.isKaboot = true;
            usBase = bidType === 'SUN' ? 44 : 25;
            themBase = 0;
        } else if (isThemKaboot) {
            // THEM won all tricks
            them.isKaboot = true;
            themBase = bidType === 'SUN' ? 44 : 25;
            usBase = 0;
        } else {
            // Normal calculation
            if (bidType === 'SUN') {
                usBase = convertSunAbnat(us.totalRaw);
                themBase = convertSunAbnat(them.totalRaw);
            } else {
                // HOKUM: Calculate winner first, loser gets remainder
                const usRounded = convertHokumAbnat(us.totalRaw);
                const themRounded = convertHokumAbnat(them.totalRaw);

                // Check if sum exceeds max (rounding can cause this)
                const sum = usRounded + themRounded;

                if (sum > maxPoints) {
                    // Winner takes calculated, loser gets remainder
                    if (usRounded >= themRounded) {
                        usBase = usRounded;
                        themBase = maxPoints - usRounded;
                    } else {
                        themBase = themRounded;
                        usBase = maxPoints - themRounded;
                    }
                } else if (sum < maxPoints) {
                    // Add remainder to winner
                    if (usRounded >= themRounded) {
                        usBase = usRounded + (maxPoints - sum);
                        themBase = themRounded;
                    } else {
                        themBase = themRounded + (maxPoints - sum);
                        usBase = usRounded;
                    }
                } else {
                    usBase = usRounded;
                    themBase = themRounded;
                }
            }
        }

        // ═══════════════════════════════════════════════════════════
        // 5. KHASARA CHECK (Buyer Win/Loss Condition)
        // ═══════════════════════════════════════════════════════════
        // The Buyer must score MORE than the threshold, otherwise they LOSE
        // and opponent takes ALL points.

        if (!isUsKaboot && !isThemKaboot && bidderTeam) {
            const bidderScore = bidderTeam === 'us' ? usBase : themBase;

            // WIN THRESHOLDS (Buyer must score MORE than this):
            // - SUN: > 13 points (meaning 14+ to win)
            // - HOKUM: > 8 points (meaning 9+ to win)
            const winThreshold = bidType === 'SUN' ? 13 : 8;

            if (bidderScore <= winThreshold) {
                // ══════════════════════════════════════
                // KHASARA! Buyer Lost!
                // Opponent takes ALL points from the pool
                // ══════════════════════════════════════
                if (bidderTeam === 'us') {
                    usBase = 0;
                    themBase = maxPoints; // 26 for SUN, 16 for HOKUM
                } else {
                    themBase = 0;
                    usBase = maxPoints;
                }
            }
        }

        // ═══════════════════════════════════════════════════════════
        // 6. APPLY BASE POINTS
        // ═══════════════════════════════════════════════════════════
        us.gamePoints = usBase;
        them.gamePoints = themBase;

        // ═══════════════════════════════════════════════════════════
        // 7. DOUBLING MULTIPLIER
        // ═══════════════════════════════════════════════════════════
        // When doubled: Winner takes (Total Points + Project Points) × Multiplier
        // Loser gets 0
        // IMPORTANT: Baloot 2 pts is NEVER multiplied - added as flat scalar at end

        if (doublingLevel > 1) {
            const totalGamePoints = us.gamePoints + them.gamePoints;

            if (us.gamePoints > them.gamePoints) {
                // US wins - takes all × multiplier
                us.gamePoints = totalGamePoints * doublingLevel;
                them.gamePoints = 0;
            } else if (them.gamePoints > us.gamePoints) {
                // THEM wins - takes all × multiplier
                them.gamePoints = totalGamePoints * doublingLevel;
                us.gamePoints = 0;
            }
            // Tie: Both keep their multiplied scores (rare)
        }

        // ═══════════════════════════════════════════════════════════
        // 8. BALOOT BONUS (Flat 2 points, NEVER multiplied)
        // ═══════════════════════════════════════════════════════════
        if (hasBalootUs) {
            us.gamePoints += 2;
        }
        if (hasBalootThem) {
            them.gamePoints += 2;
        }

        // ═══════════════════════════════════════════════════════════
        // 9. DETERMINE WINNER & RETURN
        // ═══════════════════════════════════════════════════════════
        const winner = us.gamePoints > them.gamePoints ? 'us'
            : them.gamePoints > us.gamePoints ? 'them'
                : 'tie';

        return {
            us,
            them,
            winner,
            baida: false // Baida (shutout) tracking if needed
        };
    }

    // ═══════════════════════════════════════════════════════════════
    // HELPER: Calculate Project Abnat Value
    // ═══════════════════════════════════════════════════════════════
    public static getProjectAbnatValue(type: ProjectType, bidType: 'SUN' | 'HOKUM'): number {
        const isSun = bidType === 'SUN';

        switch (type) {
            case ProjectType.FOUR_HUNDRED:
                // 4 Aces - Only valid in SUN
                return isSun ? 40 : 0;
            case ProjectType.HUNDRED:
                // 5-sequence or 4-of-a-kind (K/Q/J/10)
                return isSun ? 20 : 10;
            case ProjectType.FIFTY:
                // 4-sequence
                return isSun ? 10 : 5;
            case ProjectType.SIRA:
                // 3-sequence
                return isSun ? 4 : 2;
            case ProjectType.BALOOT:
                // K+Q of Trump - handled separately (flat 2 game points)
                return 0; // Don't add to Abnat, add directly to game points
            default:
                return 0;
        }
    }

    // ═══════════════════════════════════════════════════════════════
    // HELPER: Debug - Show calculation step by step
    // ═══════════════════════════════════════════════════════════════
    public static explainCalculation(
        usRaw: number,
        themRaw: number,
        bidType: 'SUN' | 'HOKUM',
        bidderTeam: 'us' | 'them' | null
    ): string[] {
        const steps: string[] = [];

        steps.push(`📊 Round Type: ${bidType}`);
        steps.push(`🎴 Raw Abnat - Us: ${usRaw}, Them: ${themRaw}`);

        if (bidType === 'SUN') {
            const usPoints = (usRaw * 2) / 10;
            const themPoints = (themRaw * 2) / 10;
            steps.push(`📐 SUN Formula: (Abnat × 2) / 10`);
            steps.push(`   Us: (${usRaw} × 2) / 10 = ${usPoints}`);
            steps.push(`   Them: (${themRaw} × 2) / 10 = ${themPoints}`);

            if (bidderTeam) {
                const bidderScore = bidderTeam === 'us' ? usPoints : themPoints;
                steps.push(`👤 Buyer (${bidderTeam}) scored: ${bidderScore}`);
                steps.push(`   Win threshold: > 13 points`);
                if (bidderScore <= 13) {
                    steps.push(`❌ KHASARA! Buyer lost. Opponent gets 26 points.`);
                } else {
                    steps.push(`✅ Buyer wins!`);
                }
            }
        } else {
            const usPoints = Math.floor(usRaw / 10);
            const themPoints = Math.floor(themRaw / 10);
            steps.push(`📐 HOKUM Formula: Round(Abnat / 10), 0.5 rounds DOWN`);
            steps.push(`   Us: ${usRaw} / 10 = ${(usRaw / 10).toFixed(1)} → ${usPoints}`);
            steps.push(`   Them: ${themRaw} / 10 = ${(themRaw / 10).toFixed(1)} → ${themPoints}`);

            if (bidderTeam) {
                const bidderScore = bidderTeam === 'us' ? usPoints : themPoints;
                steps.push(`👤 Buyer (${bidderTeam}) scored: ${bidderScore}`);
                steps.push(`   Win threshold: > 8 points`);
                if (bidderScore <= 8) {
                    steps.push(`❌ KHASARA! Buyer lost. Opponent gets 16 points.`);
                } else {
                    steps.push(`✅ Buyer wins!`);
                }
            }
        }

        return steps;
    }
}

File: mobile/lib/services/accounting_engine.dart
/// accounting_engine.dart — Scoring engine for Baloot AI.
///
/// Port of frontend/src/services/AccountingEngine.ts
///
/// Validated 100% against 1,095 benchmark rounds.
///
/// ## Key Constants
/// - **SUN Total**: 130 Abnat (120 cards + 10 last trick) → 26 Game Points
/// - **HOKUM Total**: 162 Abnat (152 cards + 10 last trick) → 16 Game Points
///
/// ## Scoring Formulas (Benchmark-validated)
/// - **SUN**: Floor-to-even → `divmod(abnat, 5); q + (1 if q is odd and r > 0)`
/// - **HOKUM**: Pair-based rounding with sum=16 constraint
///   - Individual: `divmod(abnat, 10); q + (1 if r > 5)`
///   - Then adjust pair if sum ≠ 16
///
/// ## Khasara (Buyer Loss)
/// - `bidder_gp < opp_gp` → khasara
/// - GP tie: compare raw abnat (doubled: doubler loses; equal raw → split)
///
/// ## Kaboot (Capot — All Tricks)
/// - SUN: 44 points
/// - HOKUM: 25 points
///
/// ## Pipeline
/// card_gp → project_gp → khasara → multiplier → baloot (last, never multiplied)
library;
import '../models/enums.dart';

/// Breakdown of a team's score for one round.
class ScoreBreakdown {
  /// Raw card points (abnat) won from tricks.
  final int rawCardPoints;

  /// Raw project points (abnat) from declared projects.
  final int projectPoints;

  /// Total raw abnat (cards + projects) before conversion.
  final int totalRaw;

  /// Final game points (GP) after conversion, khasara, multiplier, and baloot.
  int gamePoints;

  /// Whether this team won all 8 tricks (kaboot/capot).
  bool isKaboot;

  /// The doubling multiplier applied this round.
  final DoublingLevel multiplierApplied;

  ScoreBreakdown({
    required this.rawCardPoints,
    required this.projectPoints,
    required this.totalRaw,
    this.gamePoints = 0,
    this.isKaboot = false,
    this.multiplierApplied = DoublingLevel.normal,
  });

  Map<String, dynamic> toJson() => {
        'rawCardPoints': rawCardPoints,
        'projectPoints': projectPoints,
        'totalRaw': totalRaw,
        'gamePoints': gamePoints,
        'isKaboot': isKaboot,
        'multiplierApplied': multiplierApplied.value,
      };
}

/// Result of a round scoring calculation.
class RoundScoreResult {
  /// Score breakdown for "Us" (bottom + top team).
  final ScoreBreakdown us;

  /// Score breakdown for "Them" (right + left team).
  final ScoreBreakdown them;

  /// Which team won this round ('us', 'them', or 'tie').
  final String winner;

  /// Whether this round resulted in a shutout (baida).
  final bool baida;

  const RoundScoreResult({
    required this.us,
    required this.them,
    required this.winner,
    this.baida = false,
  });
}

/// Scoring engine for Baloot AI — the authoritative source for GP calculation.
///
/// All methods are static. The scoring pipeline mirrors the validated
/// benchmark formulas exactly:
///
/// 1. Convert raw abnat → base GP using mode-specific rounding.
/// 2. Check for kaboot (capot — all 8 tricks won).
/// 3. Apply khasara (buyer loss) rules with tie-breaking.
/// 4. Apply doubling multiplier.
/// 5. Add baloot bonus (flat 2 GP, NEVER multiplied).
class AccountingEngine {
  AccountingEngine._(); // Prevent instantiation

  // ═══════════════════════════════════════════════════════════════
  // CONVERSION FUNCTIONS (Abnat → Game Points)
  // ═══════════════════════════════════════════════════════════════

  /// SUN card GP: Floor-to-even rounding.
  ///
  /// Formula: `divmod(abnat, 5) → q + (1 if q is odd and r > 0)`
  /// Total pool: 130 Abnat → always sums to 26 GP.
  ///
  /// Example: 67 → 67÷5 = 13 r 2 → 13 is odd, r>0 → 14 GP
  /// Example: 63 → 63÷5 = 12 r 3 → 12 is even → 12 GP (partner gets 14)
  static int sunCardGP(int abnat) {
    final q = abnat ~/ 5;
    final r = abnat % 5;
    return q + ((q % 2 == 1 && r > 0) ? 1 : 0);
  }

  /// HOKUM individual card GP: round up if remainder > 5.
  ///
  /// Formula: `divmod(abnat, 10) → q + (1 if r > 5)`
  /// Used per-team; must apply sum=16 constraint via [hokumPairGP].
  static int hokumCardGP(int abnat) {
    final q = abnat ~/ 10;
    final r = abnat % 10;
    return q + (r > 5 ? 1 : 0);
  }

  /// HOKUM pair-based GP: ensures the pair always sums to 16.
  ///
  /// Computes individual GP for each team, then adjusts if sum ≠ 16:
  /// - sum=17: reduce the team with the larger `raw % 10` remainder.
  /// - sum=15: increase the team with the larger `raw % 10` remainder.
  /// - If remainders are equal, the team with larger raw gets adjusted.
  ///
  /// Returns `[gpA, gpB]` where A+B = 16.
  static List<int> hokumPairGP(int rawA, int rawB) {
    var gpA = hokumCardGP(rawA);
    var gpB = hokumCardGP(rawB);
    final total = gpA + gpB;

    if (total == 17) {
      final remA = rawA % 10;
      final remB = rawB % 10;
      if (remA > remB || (remA == remB && rawA >= rawB)) {
        gpA -= 1;
      } else {
        gpB -= 1;
      }
    } else if (total == 15) {
      final remA = rawA % 10;
      final remB = rawB % 10;
      if (remA > remB || (remA == remB && rawA >= rawB)) {
        gpA += 1;
      } else {
        gpB += 1;
      }
    }

    return [gpA, gpB];
  }

  // ═══════════════════════════════════════════════════════════════
  // MAIN SCORING FUNCTION
  // ═══════════════════════════════════════════════════════════════

  /// Calculates the final Game Points for a round.
  ///
  /// This is the **truth source** for scoring — matches benchmark rules exactly.
  ///
  /// Parameters:
  /// - [usRaw], [themRaw]: Raw abnat from card tricks (includes last trick bonus).
  /// - [usProjects], [themProjects]: Sum of project abnat values.
  /// - [bidType]: 'SUN' or 'HOKUM' — determines conversion formula.
  /// - [doublingLevel]: Current doubling state (NORMAL, DOUBLE, TRIPLE, etc.).
  /// - [bidderTeam]: Which team won the bid ('us' or 'them'), or null if no bid.
  /// - [hasBalootUs], [hasBalootThem]: Whether each team declared Baloot (K+Q of trump).
  ///
  /// Scoring pipeline:
  /// 1. Build initial structures with raw totals.
  /// 2. Detect kaboot (opponent scored 0 raw abnat).
  /// 3. Convert abnat → base GP using mode-specific formula.
  /// 4. Apply khasara rules (bidder loss check with tie-breaking).
  /// 5. Apply doubling multiplier (winner takes all × level).
  /// 6. Add baloot bonus (flat 2 GP, immune to multiplier).
  /// 7. Determine round winner.
  static RoundScoreResult calculateRoundResult({
    required int usRaw,
    required int themRaw,
    required int usProjects,
    required int themProjects,
    required String bidType,
    required DoublingLevel doublingLevel,
    required String? bidderTeam,
    bool hasBalootUs = false,
    bool hasBalootThem = false,
  }) {
    // ═══════════════════════════════════════════════════════════
    // 1. INITIAL STRUCTURES
    // ═══════════════════════════════════════════════════════════
    final us = ScoreBreakdown(
      rawCardPoints: usRaw,
      projectPoints: usProjects,
      totalRaw: usRaw + usProjects,
      multiplierApplied: doublingLevel,
    );

    final them = ScoreBreakdown(
      rawCardPoints: themRaw,
      projectPoints: themProjects,
      totalRaw: themRaw + themProjects,
      multiplierApplied: doublingLevel,
    );

    // ═══════════════════════════════════════════════════════════
    // 2. KABOOT DETECTION (All Tricks Won)
    // ═══════════════════════════════════════════════════════════
    final isUsKaboot = themRaw == 0 && usRaw > 0;
    final isThemKaboot = usRaw == 0 && themRaw > 0;

    // ═══════════════════════════════════════════════════════════
    // 3 & 4. BASE CALCULATION (Abnat → Game Points)
    // ═══════════════════════════════════════════════════════════
    int usBase = 0;
    int themBase = 0;

    if (isUsKaboot) {
      us.isKaboot = true;
      usBase = bidType == 'SUN' ? 44 : 25;
      themBase = 0;
    } else if (isThemKaboot) {
      them.isKaboot = true;
      themBase = bidType == 'SUN' ? 44 : 25;
      usBase = 0;
    } else {
      // Normal calculation
      if (bidType == 'SUN') {
        usBase = sunCardGP(us.totalRaw);
        themBase = sunCardGP(them.totalRaw);
      } else {
        // HOKUM: pair-based rounding with sum=16 constraint
        final pair = hokumPairGP(us.totalRaw, them.totalRaw);
        usBase = pair[0];
        themBase = pair[1];
      }
    }

    // ═══════════════════════════════════════════════════════════
  // 5. KHASARA CHECK (Benchmark-validated tie-break rules)
    // ═══════════════════════════════════════════════════════════
    if (!isUsKaboot && !isThemKaboot && bidderTeam != null) {
      final bidderScore = bidderTeam == 'us' ? usBase : themBase;
      final oppScore = bidderTeam == 'us' ? themBase : usBase;
      var khasara = false;

      if (bidderScore < oppScore) {
        khasara = true;
      } else if (bidderScore == oppScore) {
        // GP tie: compare raw abnat totals
        final bidderRaw =
            bidderTeam == 'us' ? us.totalRaw : them.totalRaw;
        final oppRaw =
            bidderTeam == 'us' ? them.totalRaw : us.totalRaw;
        final isDoubled = doublingLevel.value >= 2;

        if (isDoubled) {
          // Doubled rounds: doubler always loses the tie
          khasara = true;
        } else if (bidderRaw < oppRaw) {
          // Normal: bidder loses if raw abnat is strictly less
          khasara = true;
        }
        // Equal raw on tie → split (no khasara)
      }

      if (khasara) {
        final totalPot = usBase + themBase;
        if (bidderTeam == 'us') {
          usBase = 0;
          themBase = totalPot;
        } else {
          themBase = 0;
          usBase = totalPot;
        }
      }
    }

    // ═══════════════════════════════════════════════════════════
    // 6. APPLY BASE POINTS
    // ═══════════════════════════════════════════════════════════
    us.gamePoints = usBase;
    them.gamePoints = themBase;

    // ═══════════════════════════════════════════════════════════
    // 7. DOUBLING MULTIPLIER
    // ═══════════════════════════════════════════════════════════
    if (doublingLevel.value > 1) {
      final totalGamePoints = us.gamePoints + them.gamePoints;

      if (us.gamePoints > them.gamePoints) {
        us.gamePoints = totalGamePoints * doublingLevel.value;
        them.gamePoints = 0;
      } else if (them.gamePoints > us.gamePoints) {
        them.gamePoints = totalGamePoints * doublingLevel.value;
        us.gamePoints = 0;
      }
      // Tie: both keep their multiplied scores (rare)
    }

    // ═══════════════════════════════════════════════════════════
    // 8. BALOOT BONUS (Flat 2 points, NEVER multiplied)
    // ═══════════════════════════════════════════════════════════
    if (hasBalootUs) {
      us.gamePoints += 2;
    }
    if (hasBalootThem) {
      them.gamePoints += 2;
    }

    // ═══════════════════════════════════════════════════════════
    // 9. DETERMINE WINNER & RETURN
    // ═══════════════════════════════════════════════════════════
    final winner = us.gamePoints > them.gamePoints
        ? 'us'
        : them.gamePoints > us.gamePoints
            ? 'them'
            : 'tie';

    return RoundScoreResult(
      us: us,
      them: them,
      winner: winner,
      baida: false,
    );
  }

  // ═══════════════════════════════════════════════════════════════
  // HELPER: Calculate Project Abnat Value
  // ═══════════════════════════════════════════════════════════════

  /// Returns the raw abnat value of a project type in the given bid mode.
  ///
  /// - **400** (4 Aces): SUN=40, HOKUM=0 (invalid in HOKUM).
  /// - **100** (5-sequence or 4-of-a-kind): SUN=20, HOKUM=10.
  /// - **50** (4-sequence): SUN=10, HOKUM=5.
  /// - **Sira** (3-sequence): SUN=4, HOKUM=2.
  /// - **Baloot** (K+Q of trump): Always 0 here — handled as flat 2 GP separately.
  static int getProjectAbnatValue(ProjectType type, String bidType) {
    final isSun = bidType == 'SUN';
    switch (type) {
      case ProjectType.fourHundred:
        return isSun ? 40 : 0;
      case ProjectType.hundred:
        return isSun ? 20 : 10;
      case ProjectType.fifty:
        return isSun ? 10 : 5;
      case ProjectType.sira:
        return isSun ? 4 : 2;
      case ProjectType.baloot:
        return 0; // Flat 2 GP, not added to abnat
    }
  }

  // ═══════════════════════════════════════════════════════════════
  // DEBUG HELPER: Explain calculation step by step
  // ═══════════════════════════════════════════════════════════════

  /// Returns a step-by-step explanation of the scoring calculation.
  ///
  /// Useful for debugging and display in the score breakdown modal.
  static List<String> explainCalculation({
    required int usRaw,
    required int themRaw,
    required String bidType,
    required String? bidderTeam,
  }) {
    final steps = <String>[];

    steps.add('Round Type: $bidType');
    steps.add('Raw Abnat - Us: $usRaw, Them: $themRaw');

    if (bidType == 'SUN') {
      final usPoints = sunCardGP(usRaw);
      final themPoints = sunCardGP(themRaw);
      steps.add('SUN Formula: floor-to-even (divmod by 5)');
      steps.add('   Us: $usRaw -> $usPoints GP');
      steps.add('   Them: $themRaw -> $themPoints GP');

      if (bidderTeam != null) {
        final bidderScore = bidderTeam == 'us' ? usPoints : themPoints;
        final oppScore = bidderTeam == 'us' ? themPoints : usPoints;
        steps.add('Buyer ($bidderTeam) scored: $bidderScore vs $oppScore');
        if (bidderScore < oppScore) {
          steps.add('KHASARA! Buyer GP < opponent GP.');
        } else if (bidderScore == oppScore) {
          steps.add('GP tie -> compare raw abnat for khasara.');
        } else {
          steps.add('Buyer wins!');
        }
      }
    } else {
      final usPoints = hokumCardGP(usRaw);
      final themPoints = hokumCardGP(themRaw);
      steps.add('HOKUM Formula: pair-based rounding (sum=16)');
      steps.add('   Us: $usRaw -> $usPoints GP');
      steps.add('   Them: $themRaw -> $themPoints GP');

      if (bidderTeam != null) {
        final bidderScore = bidderTeam == 'us' ? usPoints : themPoints;
        final oppScore = bidderTeam == 'us' ? themPoints : usPoints;
        steps.add('Buyer ($bidderTeam) scored: $bidderScore vs $oppScore');
        if (bidderScore < oppScore) {
          steps.add('KHASARA! Buyer GP < opponent GP.');
        } else if (bidderScore == oppScore) {
          steps.add('GP tie -> compare raw abnat for khasara.');
        } else {
          steps.add('Buyer wins!');
        }
      }
    }

    return steps;
  }
}
